src/compositor.rs

```diff
@@ -1,73 +1,199 @@
+                Err(CompositorError(e))
+            }
         }
     }
 
-    /// Check if compositor can render a scene
-    pub fn can_render_scene(&self) -> bool {
-        unsafe {
-            let comp = * { self.0 as *mut openvr_sys::VR_IVRCompositor_FnTable };
-            comp.CanRenderScene.unwrap()() > 0
+    /// Display the supplied texture for the next frame.
+    ///
+    /// If `bounds` is None, the entire texture will be used. Lens distortion is handled by the OpenVR implementation.
+    pub fn submit(&self, eye: Eye, texture: &Texture, bounds: Option<&TextureBounds>) -> Result<(), CompositorError> {
+        use self::TextureHandle::*;
+        let texture = sys::Texture_t {
+            handle: match texture.handle {
+                Vulkan(ref x) => x as *const _ as *mut _,
+            },
+            eType: match texture.handle {
+                Vulkan(_) => sys::ETextureType_ETextureType_TextureType_Vulkan,
+            },
+            eColorSpace: texture.color_space as sys::EColorSpace,
+        };
+        let e = unsafe {
+            (self.0.Submit.unwrap())(eye as sys::EVREye,
+                                     &texture as *const _ as *mut _,
+                                     bounds.map(|x| x as *const _ as *mut TextureBounds as *mut _).unwrap_or(ptr::null_mut()),
+                                     sys::EVRSubmitFlags_EVRSubmitFlags_Submit_Default)
+        };
+        if e == sys::EVRCompositorError_EVRCompositorError_VRCompositorError_None {
+            Ok(())
+        } else {
+            Err(CompositorError(e))
         }
     }
 
-    /// Submits an opengl framebuffer as an eye to the render
-    pub fn submit(&self, eye: Eye, texture: usize, bounds: TextureBounds) {
-        let mut b = bounds.to_raw();
-        let e = eye.to_raw();
+    pub fn post_present_handoff(&self) {
+        unsafe { (self.0.PostPresentHandoff.unwrap())() };
+    }
+}
 
-        unsafe {
-            use std;
-
-            let comp = * { self.0 as *mut openvr_sys::VR_IVRCompositor_FnTable };
-            let mut t = openvr_sys::Texture_t {
-                eType: EGraphicsAPIConvention_API_OpenGL,
-                eColorSpace: EColorSpace_ColorSpace_Auto,
-                handle: texture as *mut std::os::raw::c_void,
-            };
-
-            comp.Submit.unwrap()(
-                e,
-                &mut t,
-                &mut b as *mut openvr_sys::VRTextureBounds_t,
-                EVRSubmitFlags_Submit_GlRenderBuffer
-            );
-        }
+#[derive(Debug, Copy, Clone)]
+pub struct WaitPoses {
+    /// Predicted to the point they will be at the upcoming frame.
+    pub render: TrackedDevicePoses,
+    /// Predicted to the point they will be at the frame after the upcoming frame, for use in game logic.
+    pub game: TrackedDevicePoses,
+}
+
+pub use sys::VkPhysicalDevice_T;
+pub use sys::VkDevice_T;
+pub use sys::VkInstance_T;
+pub use sys::VkQueue_T;
+
+#[derive(Debug, Copy, Clone)]
+pub struct Texture {
```

Comment:
Reminder:
ToDo: I'd prefer it to split the compositor into a module with multiple submodules for texture, etc.
```
+ compositor
    + mod.rs <- general logic
    + texture<- texture related stuff
           + mod.rs
           + vulkan.rs
           + opengl.rs
    + error.rs <- error enum ?maybe not?
```